fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabaseRunnerTest.java

```diff
@@ -355,4 +364,65 @@ public void close() throws Exception {
         Thread.sleep(150);
         assertEquals(currentIteration, iteration.get(), "Should have stopped running");
     }
+
+    @Test
+    void testRestoreMdc() {
+        Executor oldExecutor = FDBDatabaseFactory.instance().getExecutor();
+        try {
+            ThreadContext.clearAll();
+            ThreadContext.put("outer", "Echidna");
+            Map<String, String> outer = ThreadContext.getContext();
+            final ImmutableMap<String, String> restored = ImmutableMap.of("restored", "Platypus");
+
+            FDBDatabaseFactory.instance().setExecutor(new FDBRecordContext.ContextRestoringExecutor(
+                    new ForkJoinPool(2), ImmutableMap.of("executor", "Water Bear")));
+            AtomicInteger attempts = new AtomicInteger(0);
+            final FDBDatabaseRunner runner = new FDBDatabaseRunner(database, null, restored);
+            List<Map<String, String>> threadContexts = new Vector<>();
+            Consumer<String> saveThreadContext =
+                    name -> threadContexts.add(threadContextPlus(name, attempts.get(), ThreadContext.getContext()));
+            final String runnerRunAsyncName = "runner runAsync";
+            final String supplyAsyncName = "supplyAsync";
+            final String handleName = "handle";
+            assertNull(runner.runAsync(recordContext -> {
+                saveThreadContext.accept(runnerRunAsyncName);
+                return CompletableFuture.supplyAsync(() -> {
+                    saveThreadContext.accept(supplyAsyncName);
+                    if (attempts.getAndIncrement() == 0) {
+                        throw new RecordCoreRetriableTransactionException("Retriable and lessener",
+                                new FDBException("not_committed", 1020));
+                    } else {
+                        return null;
+                    }
+                }, recordContext.getExecutor());
+            }).handle((result, exception) -> {
+                saveThreadContext.accept(handleName);
+                return exception;
+            }).join());
+            List<Map<String, String>> expected = ImmutableList.of(
+                    // first attempt:
+                    // it is known behavior that the first will be run in the current context
+                    threadContextPlus(runnerRunAsyncName, 0, outer),
```

Comment:
So, this will be because the first iteration of the loop occurs in the current thread rather than on any thread spawned by the executors?